File: Xaas/Python_이상탐지 알고리즘/k_means.py
import numpy as np
from disteusq import disteusq


def k_means(d, k, x0=None):
    
    n, _ = d.shape

    if x0 is None:
        x = d[np.random.choice(n, k, replace=False), :]
        
    else:
        x = x0

    y = x + 1
   
    iter = 0
    while not np.all(x == y):
        iter += 1
        
        z = disteusq(x=d, y=x, mode='x')
    
        m = np.min(z, axis=1)
        j = np.argmin(z, axis=1)
      
        y = x.copy()

        for i in range(k):
            
            s = j == i
            if np.any(s):
                x[i, :] = np.mean(d[s, :], axis=0)
            else:
                q = np.where(m != 0)[0]
                if len(q) == 0:
                    break
                r = np.random.choice(q)
                x[i, :] = d[r, :]
                m[r] = 0

                y = x.copy() + 1

    esq = np.mean(m)
    
    return x, esq, j

File: Xaas/Comparative_study/0_PCA_SS/k_means.m
function [x,esq,j] = k_means(d,k,x0)
% K-means 

[n,p] = size(d);
if nargin<3
   x = d(ceil(rand(1,k)*n),:);
else
   x=x0;
end
y = x+1;
iter = 0;
while any(x(:) ~= y(:))
   iter = iter + 1;
   z = disteusq(d,x,'x');
   [m,j] = min(z,[],2);
   y = x;
   for i=1:k
      s = j==i;
      if any(s)
         x(i,:) = mean(d(s,:),1);
      else
         q=find(m~=0);
         if isempty(q) break; end
         r=q(ceil(rand*length(q)));
         x(i,:) = d(r,:);
         m(r)=0;
         y=x+1;
      end
   end
end
esq=mean(m,1);